Objects/longobject.c

```diff
@@ -4491,38 +4493,57 @@ long_rshift1(PyLongObject *a, Py_ssize_t wordshift, digit remshift)
 {
     PyLongObject *z = NULL;
     Py_ssize_t newsize, hishift, i, j;
-    digit lomask, himask;
+    digit lomask, himask, omitmark;
 
-    if (Py_SIZE(a) < 0) {
-        /* Right shifting negative numbers is harder */
-        PyLongObject *a1, *a2;
-        a1 = (PyLongObject *) long_invert(a);
-        if (a1 == NULL)
-            return NULL;
-        a2 = (PyLongObject *) long_rshift1(a1, wordshift, remshift);
-        Py_DECREF(a1);
-        if (a2 == NULL)
-            return NULL;
-        z = (PyLongObject *) long_invert(a2);
-        Py_DECREF(a2);
+    if (IS_MEDIUM_VALUE(a)) {
+        stwodigits x;
+        if (wordshift > 0) {
+            return get_small_int(-(Py_SIZE(a) < 0));
+        }
+        x = Py_ARITHMETIC_RIGHT_SHIFT(stwodigits, medium_value(a), remshift);
```

Comment:
Suggestion that may remove a branch on common compilers:

```suggestion
        int shift = wordshift == 0 ? remshift : PyLong_SHIFT;
        x = Py_ARITHMETIC_RIGHT_SHIFT(stwodigits, medium_value(a), shift);
```